vllm/platforms/cuda.py

```diff
@@ -203,18 +203,29 @@ def get_current_memory_usage(cls,
         return torch.cuda.max_memory_allocated(device)
 
     @classmethod
-    def get_vit_attn_backend(cls, support_fa: bool = False) -> _Backend:
-        if cls.has_device_capability(80) and support_fa:
-            from transformers.utils import is_flash_attn_2_available
-            if is_flash_attn_2_available():
-                return _Backend.FLASH_ATTN
-            logger.warning_once(
-                "Current `vllm-flash-attn` has a bug inside vision "
-                "module, so we use xformers backend instead. You can "
-                "run `pip install flash-attn` to use flash-attention "
-                "backend.")
+    def get_vit_attn_backend(cls, head_size: int,
+                             dtype: torch.dtype) -> tuple[_Backend, bool]:
+        if dtype not in (torch.float16, torch.bfloat16):
+            return _Backend.XFORMERS, False
+
+        if cls.has_device_capability(80):
+            if head_size % 32 == 0:
+                # Use vllm-flash-attn
+                return _Backend.FLASH_ATTN, False
+            if head_size % 32 != 0:
+                from transformers.utils import is_flash_attn_2_available
+                if is_flash_attn_2_available():
+                    # Use upstream FA
+                    return _Backend.FLASH_ATTN, True
+                else:
+                    # Fallback to XFORMERS
+                    logger.warning_once(
+                        "Using xformers for ViT attention backend. "
+                        "To use flash attention for ViT"
+                        "please install flash_attn")
+                    return _Backend.XFORMERS, False
```

Comment:
```suggestion
        if cls.has_device_capability(80):
            is_default_fa_supported = is_attn_backend_supported(
                        FLASH_ATTN_V1, head_size, dtype,
                        allow_import_error=False)
            is_upstream_fa_supported = is_flash_attn_2_available()
            if is_default_fa_supported:
                return _Backend.FLASH_ATTN, False
            elif is_upstream_fa_supported:
                return _Backend.FLASH_ATTN, True
            return _Backend.XFORMERS, False
```

`head_size % 32 == 0` maybe not a robust solution, because fork FA's support head_size might change in the future. I think we can use `is_attn_backend_supported` here.

Reply:
Updated.